File: 0001~0500/0060. Permutation Sequence/main.py
"""
Source: https://rb.gy/7fju6
Date: 2023/8/1
Skill:
Ref:
Runtime: 33 ms, faster than 98.95%
Memory Usage: 16.33 MB, less than 54.07%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        def get_permutation_num(digit_num):
            num = 1
            for i in range(2, digit_num + 1):
                num *= i
            return num

        res, visited = "", [0 for _ in range(n)]
        for i in range(n):
            per_num = get_permutation_num(n - i - 1)
            for j in range(n):
                if visited[j]:
                    continue
                if per_num < k:
                    k -= per_num
                else:
                    visited[j] = 1
                    res += str(j + 1)
                    break

        return res


if __name__ == "__main__":
    s = Solution()
    n, k = 3, 3
    res = s.getPermutation(n, k)

File: 0001~0500/0060. Permutation Sequence/main.cpp
/**
 * Source: ibit.ly/11lhZ
 * Date: 2023/11/17
 * Skill:
 * Runtime: 173 ms, faster than 8.99% of C++ online submissions
 * Memory Usage: 6.40 MB, less than 74.91% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */


#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;


class Solution {
public:
    string getPermutation(int n, int k) {
        vector<int> nums;
        for (int i = 1; i <= n; i++) nums.push_back(i);
        k -= 1;
        for (int i = 0; i < k; i++) nextPermutation(nums);
        string res;
        for (int i = 0; i < n; i++) res.push_back('0' + nums[i]);
        return res;
    }
    void nextPermutation(vector<int>& nums) {
        int n = nums.size(), pos = nums.size() - 1;
        while (pos > 0 && nums[pos] <= nums[pos - 1]) pos--;
        if (pos == 0) {
            int end = n - 1;
            while (pos < end) swap(nums[pos++], nums[end--]);
        } else {
            int a = n - 1;
            while (nums[a] <= nums[pos - 1]) a--;
            swap(nums[a], nums[pos - 1]);
            a = n - 1;
            while (pos < a) swap(nums[pos++], nums[a--]);
        }
    }
};


int main() {
    Solution s;
}